examples/include/crypto-accelerator.p4

```diff
@@ -0,0 +1,136 @@
+///     +------------------+--------------------------+-----------+-----+
+///     | Headers not to   | Encryption protocol      | cleartext | ICV |
+///     | Encrypted        | headers (E.g Esp, Esp-IV)| Payload   |     |
+///     +------------------+--  ----------------------+-----------+-----+
+///     Results: Success, Auth Failure, Hardware Error
+///
+extern crypto_accelerator {
+    /// constructor
+    /// Some methods provided in this object may be specific to an algorithm used.
+    /// Compiler may be able to check and warn/error when incorrect methods are used
+    crypto_accelerator(crypto_algorithm_e algo);
+
+
+    // security association index for this security session
+    // Some implementations do not need it.. in that case this method should result in no-op
+    void set_sa_index<T>(in T sa_index);
+
+    // Set the initialization data based on protocol used. E.g. salt, random number/ counter for ipsec
+    void set_iv<T>(in T iv);
+    void set_key<T,S>(in T key, in S key_size);   // 128, 192, 256
+
+    // The format of the auth data is not specified/mandated by this object definition
+    // If it is part of the packet, it can be specified using offset/len mothods below
+    void set_auth_data_offset<T>(in T offset);
+    void set_auth_data_len<T>(in T len);
+
+    // Alternatively: Following API can be used to consturct the auth_data and
+    // provide it to the engine.
+    void add_auth_data<H>(in H auth_data);
+
+    // Auth trailer aka ICV is added by the engine after doing encryption operation
+    // Specify icv location - when a wire protocol wants to add ICV in a specific location (e.g. AH)
+    // The following apis can be used to specify the location of ICV in the packet
+    // A special offset indicates ICV is after the payload
+    void set_icv_offset<T>(in T offset);
+    void set_icv_len<L>(in L len);
+
+    // setup payload to be encrypted/decrypted
+    void set_payload_offset<T>(in T offset);
+    void set_payload_len<T>(in T len);
+    
+    // crypto accelerator runs at the end of the pipeline (after deparser), the following
+    // methods will enable the accelerator to run encrypt/decrypt operations
+    // enable_auth flag enables authentication check for decrypt. For encrypt operation,
+    // auth data computed, is added to specified icv_offset/len
+    void enable_encrypt<T>(in T enable_auth);
+    void enable_decrypt<T>(in T enable_auth);
+
+    // disable crypto engine. Between enable and disable methods,
+    // whichever method is called last overrides the previous calls
+    void disable();
+
+    crypto_results_e get_results();       // get results of the previous operation
+
+    // set_error_action() indicates behavior on encoutering an error
+    // Default action is NO_ACTION i.e. report error via get_results()
+    // Certain actions may need action parameters, e.g send_to_port
+    void set_error_action<T>(crypto_error_action_e error_action, in T action_param);
```

Comment:
call it drop_on_error() remove other options. Name set_error_behavior().. or on_error() ? crypto_error_action_e -> crypto_error_behavior_e

Reply:
This method was added as a result of discussion to see if it makes sense. I am removing it for the following reasons - 
1. As mentioned in the code comments, it is preferable to keep the accelerator agnostic to packet redirection (send to  port), packet drop type of functions that are not related to crypto operation. 
2. Also many existing targets do not support this functionality. 
3. Any targets existing or future that support it, can add a target specific method for this.